xdot/ui/elements.py

```diff
@@ -532,12 +533,80 @@ def is_inside(self, x, y):
         return False
 
     def get_jump(self, x, y):
-        if self.is_inside_begin(x, y):
-            return Jump(self, self.dst.x, self.dst.y, highlight=set([self, self.dst]))
-        if self.is_inside_end(x, y):
-            return Jump(self, self.src.x, self.src.y, highlight=set([self, self.src]))
+        for shape in self.shapes:
+            x1, y1, x2, y2 = shape.bounding
+            if (x1 - self.RADIUS) <= x and x <= (x2 + self.RADIUS) and (y1 - self.RADIUS) <= y and y <= (y2 + self.RADIUS):
+                break
+
+        else:
+            return None
+
+        if self._get_distance(x, y) <= self.RADIUS**2:
+            return Jump(self, self.src.x, self.src.y)
+
         return None
 
+    def _get_distance(self, x, y):
+        result = float('inf')
+
+        for shape in self.shapes:
+            if not isinstance(shape, BezierShape):
+                continue
+
+            points_iter = iter(shape.points)
+            x0, y0 = next(points_iter)
+
+            while True:
+                try:
+                    x1, y1 = next(points_iter)
+
+                except StopIteration:
+                    break
+
+                x2, y2 = next(points_iter)
+                x3, y3 = next(points_iter)
+
+                _e1 = -5
+                _e2 = (x0 - 3 * x1 + 3 * x2 - x3)
+                _e3 = (y0 - 3 * y1 + 3 * y2 - y3)
+                _e4 = 2 * x1
+                _e5 = 2 * y1
+                _e6 = x0**2
+                _e7 = y0**2
+                _e8 = -2
+                _e9 = (x0 - _e4 + x2)
+                _e10 = (y0 - _e5 + y2)
+                _e11 = 2 * x0
+                _e12 = 2 * y0
+                _e13 = 5 * _e6
+                _e14 = 5 * _e7
+                _e15 = x1**2
+                _e16 = y1**2
+                coefficients = [
+                    (x - x0) * (x0 - x1) + (y - y0) * (y0 - y1),
+                    _e13 + 3 * _e15 + _e11 * (_e1 * x1 + x2) - 2 * x * _e9 + _e14 + 3 * _e16 + _e12 * (_e1 * y1 + y2) - 2 * y * _e10,
+                    -10 * _e6 + 9 * x1 * (_e8 * x1 + x2) + x * _e2 + x0 * (30 * x1 - 12 * x2 + x3) - 10 * _e7 + 9 * y1 * (_e8 * y1 + y2) + y * _e3 + y0 * (30 * y1 - 12 * y2 + y3),
+                    2 * (_e13 + 18 * _e15 + 3 * x2**2 + _e4 * (-9 * x2 + x3) - _e11 * (10 * x1 - 6 * x2 + x3) + _e14 + 3 * (6 * _e16 - 6 * y1 * y2 + y2**2) + _e5 * y3 - _e12 * (10 * y1 - 6 * y2 + y3)),
+                    _e1 * _e9 * _e2 - 5 * _e10 * _e3,
+                    _e2**2 + _e3**2
+                ]
+                coefficients.reverse()
+
+                for t in numpy.roots(coefficients):
```

Comment:
`numpy` is a new dependency.  It's not a big deal, as it's so common, but still, I wonder if we could just avoid it, just for this.

Reply:
The polynomial is of degree 5 and even Mathematica (ToRadicals) can't give a closed form solution. Therefore the only way I see is to do some kind of search, either with precomputed points as in the first version of this PR, or on demand. I guess with mostly straight lines or slightly bent curves this alternative could work just fine.

Reply:
Fair enough.  `numpy` is such a common dependency, and widely available and easily installable (far more than for example GTK), I don't have misgivings depending on it.  I just wanted to make sure it wasn't easily avoidable.

You'll need to update `.travis.yml` so the tests can run, like:

```diff
diff --git a/.travis.yml b/.travis.yml
index f49b066..7626166 100644
--- a/.travis.yml
+++ b/.travis.yml
@@ -9,6 +9,7 @@ addons:
     - gir1.2-gtk-3.0
     - python3-gi
     - python3-gi-cairo
+    - python3-numpy
     - graphviz
     - xvfb
     - twine
```